### PR TITLE
sys: next highest power-of-two (nhpot) utilities

### PR DESCRIPTION
**Pull request by cfriedt**
There are scenarios at both compile time and runtime where it is useful to know the next highest power of two.

Provide both macros and static inline functions for to fill those respective gaps as well as 64-bit variants. Additionally, the static inline function is `constexpr` to facilitate compile-time C++ optimizations.

Each of the API calls behave the same in that the return the next highest power of two.

E.g.
```
NHPOT(0) => 1
NHPOT(1) => 1
NHPOT(2) => 2
NHPOT(3) => 4
NHPOT(4) => 4
NHPOT(5) => 8
..
NHPOT64(BIT64(63)) => BIT64(63)
NHPOT64(BIT64(63)+1) => 0
```

### CODE REVIEW

**Comment by cfriedt** [comment]
I would be happy to regartet this for `<zephyr/sys/util.h>` if that's preferable, merging tests into `tests/unti/util/`

**Comment by JordanYates** [comment]
I have similar macros (calculating log2(x) instead of the next power of 2) in my tree, but implemented using compiler builtins:
```
/* Number of bits that make up a type */
#define _NUM_BITS(type) (8 * sizeof(type))

/**
 * @brief Simple macro which returns true if the supplied value is a power of 2
 *
 * @param x Value to check if power of 2
 *
 * @retval True if x is a power of 2
 */
#define IS_POWER_2(x) ((x) && !((x) & ((x) - 1)))

/** @brief Compute the log2 of an unsigned integer
 *
 * @param x value to compute logarithm of (positive only)
 *
 * @returns log2(x)
 */
#define U_LOG_2(x) _Generic((x), unsigned char : (_NUM_BITS(unsigned int) - __builtin_clz(x) - 1), \
			    unsigned short : (_NUM_BITS(unsigned int) - __builtin_clz(x) - 1),	   \
			    unsigned int : (_NUM_BITS(unsigned int) - __builtin_clz(x) - 1),	   \
			    unsigned long : (_NUM_BITS(unsigned long) - __builtin_clzl(x) - 1),	   \
			    unsigned long long : (_NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1))
```
Extending them a little to get NHPOT
```
#define NHPOT(x) (1 << (U_LOG_2(x) + !IS_POWER_2(x)))
```
And then compiling for `qemu_cortex_m3`:
```
uint32_t builtin_nhtop(uint32_t p)
{
	return NHPOT(p);
}

uint32_t or_nhpot(uint32_t val)
{
	uint32_t val2 = val;

	val2--;
	val2 |= val2 >> 1;
	val2 |= val2 >> 2;
	val2 |= val2 >> 4;
	val2 |= val2 >> 8;
	val2 |= val2 >> 16;
	val2++;

	return (0 == val) * 1 + (0 != val) * val2;
}
```
Gives the following assembly for each:
```
00007628 <builtin_nhtop>:
    7628:	b160      	cbz	r0, 7644 <builtin_nhtop+0x1c>
    762a:	1e43      	subs	r3, r0, #1
    762c:	4203      	tst	r3, r0
    762e:	bf14      	ite	ne
    7630:	2301      	movne	r3, #1
    7632:	2300      	moveq	r3, #0
    7634:	fab0 f080 	clz	r0, r0
    7638:	331f      	adds	r3, #31
    763a:	1a18      	subs	r0, r3, r0
    763c:	2301      	movs	r3, #1
    763e:	fa03 f000 	lsl.w	r0, r3, r0
    7642:	4770      	bx	lr
    7644:	2301      	movs	r3, #1
    7646:	e7f5      	b.n	7634 <builtin_nhtop+0xc>

00007648 <or_nhpot>:
    7648:	2800      	cmp	r0, #0
    764a:	f100 33ff 	add.w	r3, r0, #4294967295	; 0xffffffff
    764e:	bf14      	ite	ne
    7650:	2001      	movne	r0, #1
    7652:	2000      	moveq	r0, #0
    7654:	ea43 0353 	orr.w	r3, r3, r3, lsr #1
    7658:	ea43 0393 	orr.w	r3, r3, r3, lsr #2
    765c:	ea43 1313 	orr.w	r3, r3, r3, lsr #4
    7660:	ea43 2313 	orr.w	r3, r3, r3, lsr #8
    7664:	ea43 4313 	orr.w	r3, r3, r3, lsr #16
    7668:	fb03 0000 	mla	r0, r3, r0, r0
    766c:	bf0c      	ite	eq
    766e:	2301      	moveq	r3, #1
    7670:	2300      	movne	r3, #0
    7672:	4418      	add	r0, r3
    7674:	4770      	bx	lr

 .text.builtin_nhtop
                0x0000000000007628       0x20 app/libapp.a(main.c.obj)
                0x0000000000007628                builtin_nhtop
 .text.or_nhpot
                0x0000000000007648       0x2e app/libapp.a(main.c.obj)
                0x0000000000007648                or_nhpot
```

**Comment by cfriedt** [comment]
> I have similar macros (calculating log2(x) instead of the next power of 2) in my tree, but implemented using compiler builtins:

is `__builtin_clz()` capable of doing things at compile time? I.e. can you make a static const array with your `NHPOT` used to specify array dimensions?

File: tests/unit/nhpot/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Copyright 2023 Meta
+# SPDX-License-Identifier: Apache-2.0
+
+tests:
+  tags: pot
```
Comment (cfriedt) [inline tests/unit/nhpot/testcase.yaml:5]: Should be updated to `nhpot`

**Comment by cfriedt** [comment]
> > I have similar macros (calculating log2(x) instead of the next power of 2) in my tree, but implemented using compiler builtins:
> 
> @JordanYates - is `__builtin_clz()` compile-time const? I.e. can you make a static const array with your `NHPOT` used to specify array dimensions?

Looks like it can be used. The 2 issues I see right now are
1. Zephyr does not (yet) officially support C11 Generics
2. I get a gazillion compiler warnings because there is no specification for `int` 

I would be fine if this gets patched later too..

**Comment by cfriedt** [comment]
Closed, as it's now part of https://github.com/zephyrproject-rtos/zephyr/pull/53405

**Comment by JordanYates** [comment]
Just to finalise the discussion here, it is compile time const. This works fine:
```
int array[NHPOT(7)] = { 0 };
```
An `int` variant in the generic can easily be added, I only did unsigned because I wasn't using constants.

**Pull request closed without merging**